### ISSUE TITLE
[Enhancement] Allow Secure Storage to retrieve keys even after uninstalling app or clearing data

### ISSUE THREAD

**Issue by mina5500** [xamarin/Essentials#1346]
## Summary
Allowing Secure Storage to retrieve keys even after uninstalling app or clearing data to save unique identifier per device.

Since now with new android/ios security/privacy enhancements, its nearly impossible to read H/W components (imei,serial,mac address ...etc. )

However we can use keystore (android) and keychain (ios) in order to save a GUID for each device so that it can be saved in backend(database) and fetched for comparison & as far as i searched they are not affected after uninstalling app or clearing data.

The xamarin essentials secure storage depends on app preferences along with keystore in android which gets removed when unistalling or clearing app data

I couldn't test yet in iOS but some posts say that saved data using secure storage remains even after unistalling then installing app.

It would be great to have this feature by passing a new parameter to existing SetAsync/GetAsync or creating a new one.

## Intended Use Case
A lot of mobile apps now have a need to restrict user access to be one user per device , in order to have that the device has to marked with a unique identifier.
in my app i need only one username per device, any trial to log on with another user the username will be deactivated on this device, also it will help for removing invalid firebase installation ids that expires due to uninstalling apps.

**Comment by mattleibow** [comment]
I am not sure if this is possible... Uninstalling an up usually cleans up things and there is no way for one app to look into another app.

However, this might be solvable by tracking the user on your own server/db. For example, when a user logs in, it looks for some GUID on the server. If there is one, then the user has to either log out of the previous device (which will clear the GUID on the server) or invalidate the old GUID. Then, this GUID is passed on the requests to APIs and the server will reject a bad user/GUID pair.

Basically, instead of tracking all the users' devices, track the user. 

Does this make sense?

**Comment by mina5500** [comment]
I already save the user devices on db
But the user has the right to login with his account on as many devices as he wants but cant use muliple users on same device
That why i need to identify the device with unique id that is not affected by uninstalling apps.... by the way i found out that keys saved by secure storage in ios are not removed when unistalling apps (not using prefernces just keychain) .... the issue is in android only

**Comment by mina5500** [comment]
Any updates

**Comment by mina5500** [comment]
@jamesmontemagno  can u please help 😊?

**Comment by jamesmontemagno** [comment]
If the user syncs their keychain and syncs their android preference then it would back up automatically and even when the user uninstalls the app. Possible, but you can't guarantee this at all or how Android/iOS will actually persist data. This isn't a feature that we can support beyond what the underlying platform does today.

**Comment by jamesmontemagno** [comment]
My recommendation if this is essential to your app is roll your own implementation to exactly what you need.

**Comment by mina5500** [comment]
Thanks @jamesmontemagno  , but do you have any idea how to identify the device with any uniqueness stay persisted even after uninstalling app or clearing app data

**Comment by jamesmontemagno** [comment]
Not that I know of besides storing a GUID. iOS does not really allow getting device info and Android is not consistent -> https://montemagno.com/unique-device-id-for-mobile-apps/

**Comment by VladislavAntonyuk** [comment]
During login you can send username/password and device Id, and compare it with the data in your server dB. If user entered correct dB and device id is not found, it’s a new device.

**Comment by mina5500** [comment]
> During login you can send username/password and device Id, and compare it with the data in your server dB. If user entered correct dB and device id is not found, it’s a new device. And opposite, if device id exists in the dB and username is different, you should not allow user to login
> 
> You shouldn’t have any checks on the client side

Device ID gets changed every time you install/uninstall application.

**Comment by VladislavAntonyuk** [comment]
> > During login you can send username/password and device Id, and compare it with the data in your server dB. If user entered correct dB and device id is not found, it’s a new device. And opposite, if device id exists in the dB and username is different, you should not allow user to login
> > You shouldn’t have any checks on the client side
> 
> Device ID gets changed every time you install/uninstall application.

You can use any identifier, which is not changed after reinstalling the app

**Comment by mina5500** [comment]
> > > During login you can send username/password and device Id, and compare it with the data in your server dB. If user entered correct dB and device id is not found, it’s a new device. And opposite, if device id exists in the dB and username is different, you should not allow user to login
> > > You shouldn’t have any checks on the client side
> > 
> > Device ID gets changed every time you install/uninstall application.
> 
> You can use any identifier, which is not changed after reinstalling the app

Actually.... we don't mind logging same user on different  devices .... but not multiple users on same device... that's why we need a persistent unique identifier for each device that is not removed or changed on reinstalling our app

**Comment by mattleibow** [comment]
I'm closing this as this is not something that directly pertains to Xamarin.Essentials. We can't do too much more in many platforms due to privacy and security limitations.

**Comment by mina5500** [comment]
actually we used secure storage for ios and old nugget DeviceInfo by James for Android and it is working fine (Persistent Unique Identifier even after app removal for same publisher) till latest iOS 14 and Android 11
